src/buildercore/threadbare/state.py

```diff
@@ -0,0 +1,130 @@
+import copy
+import contextlib
+
+CLEANUP_KEY = "_cleanup"
+
+
+class LockableDict(dict):
+    def __init__(self, *args, **kwargs):
+        dict.__init__(self, *args, **kwargs)
+        self.read_only = False
+
+    def update(self, new_dict):
+        if self.read_only:
+            raise ValueError(
+                "dictionary is locked attempting to `update` with %r" % new_dict
+            )
+        dict.update(self, new_dict)
+
+    def __setitem__(self, key, val):
+        # I suspect multiprocessing isn't copying the custom 'read_only' attribute back
+        # from the child process results. be aware of this weirdness
```

Comment:
I am not sure I know all the stack of libraries that provides multiple processes here, but it's usually safe to say that once you have multiple processes variables diverge and are not in communication with each other. Whatever communication channel is there would be built on stdin/stdout?

Reply:
Assuming the results are encoded on stdout somehow.

Reply:
> Whatever communication channel is there would be built on stdin/stdout?

in this case, it's Python's `multiprocessing` that is doing the encoding of data and passing it to child processes [here](https://github.com/elifesciences/threadbare/blob/1.0-prep/threadbare/execute.py#L30-L57). I was having unexpected behaviour here as new dicts should have been locked and empty, instead they were empty and unlocked, which to me meant that the the `state` module wasn't being loaded properly somehow as the `env` is initialised and locked in the global scope. 

I ended up resetting the state before executing the `worker_func` and this seems to work fine.